### PR TITLE
Implement methods of ProtectedBranchClient

### PR DESCRIPTION
**Pull request by PrashanthiRamesh**

### CODE REVIEW

File: NGitLab.Mock/Clients/ProtectedBranchClient.cs [hunk 1]
```diff
@@ -13,24 +14,41 @@ public ProtectedBranchClient(ClientContext context, int projectId)
...
         {
             throw new NotImplementedException();
         }
 
         public void UnprotectBranch(string branchName)
         {
-            throw new NotImplementedException();
+            using (Context.BeginOperationScope())
+            {
+                var project = GetProject(_projectId, ProjectPermission.View);
```
Comment (meziantou) [inline NGitLab.Mock/Clients/ProtectedBranchClient.cs]: You are going to edit the project, let's validate permissions

```suggestion
                var project = GetProject(_projectId, ProjectPermission.Edit);
```

File: NGitLab.Mock/Clients/ProtectedBranchClient.cs [hunk 2]
```diff
@@ -13,24 +14,41 @@ public ProtectedBranchClient(ClientContext context, int projectId)
...
+            }
         }
 
-        public ProtectedBranch[] GetProtectedBranches(string search = null)
+        public Models.ProtectedBranch[] GetProtectedBranches(string search)
         {
-            throw new NotImplementedException();
+            using (Context.BeginOperationScope())
+            {
+                var project = GetProject(_projectId, ProjectPermission.View);
```
Comment (meziantou) [inline NGitLab.Mock/Clients/ProtectedBranchClient.cs]:
```suggestion
                var project = GetProject(_projectId, ProjectPermission.Edit);
```

File: NGitLab.Mock/Clients/ProtectedBranchClient.cs [hunk 3]
```diff
@@ -13,24 +14,41 @@ public ProtectedBranchClient(ClientContext context, int projectId)
             _projectId = projectId;
         }
 
-        public ProtectedBranch GetProtectedBranch(string branchName)
+        public Models.ProtectedBranch GetProtectedBranch(string branchName)
         {
-            throw new NotImplementedException();
+            using (Context.BeginOperationScope())
+            {
+                var project = GetProject(_projectId, ProjectPermission.View);
```
Comment (meziantou) [inline NGitLab.Mock/Clients/ProtectedBranchClient.cs]:
```suggestion
                var project = GetProject(_projectId, ProjectPermission.Edit);
```

File: NGitLab.Mock/Clients/ProtectedBranchClient.cs [hunk 4]
```diff
@@ -13,24 +14,41 @@ public ProtectedBranchClient(ClientContext context, int projectId)
...
         }
 
-        public ProtectedBranch[] GetProtectedBranches(string search = null)
+        public Models.ProtectedBranch[] GetProtectedBranches(string search)
         {
-            throw new NotImplementedException();
+            using (Context.BeginOperationScope())
+            {
+                var project = GetProject(_projectId, ProjectPermission.View);
+                return project.ProtectedBranches.Select(b => b.ToProtectedBranchClient()).ToArray();
```
Comment (meziantou) [inline NGitLab.Mock/Clients/ProtectedBranchClient.cs]:
```suggestion
                return project.ProtectedBranches
                           .Where(b => b.Name.Contains(search ?? "", stringComparisong.Ordinal))
                           .Select(b => b.ToProtectedBranchClient())
                           .ToArray();
```

File: NGitLab.Mock/Clients/ProtectedBranchClient.cs [hunk 5]
```diff
@@ -13,24 +14,41 @@ public ProtectedBranchClient(ClientContext context, int projectId)
...
             throw new NotImplementedException();
         }
 
         public void UnprotectBranch(string branchName)
         {
-            throw new NotImplementedException();
+            using (Context.BeginOperationScope())
+            {
+                var project = GetProject(_projectId, ProjectPermission.View);
+                var protectedBranches = project.ProtectedBranches.Select(b => b.ToProtectedBranchClient()).ToList();
```
Comment (meziantou) [inline NGitLab.Mock/Clients/ProtectedBranchClient.cs]: I think this is useless. You need to edit `project.ProtectedBranches` directly. Something like

````
var protectedBranch = project.ProtectedBranches.FirstOrDefault(b => b.Name == branchName);
if(protectedBranch != null)
{
    project.ProtectedBranches.Remove(protectedBranch);
}
````

File: NGitLab.Mock/Clients/ProtectedBranchClient.cs [hunk 6]
```diff
@@ -13,24 +14,41 @@ public ProtectedBranchClient(ClientContext context, int projectId)
...
+            {
+                var project = GetProject(_projectId, ProjectPermission.View);
+                return project.ProtectedBranches.Select(b => b.ToProtectedBranchClient()).ToArray();
+            }
         }
 
-        public ProtectedBranch ProtectBranch(BranchProtect branchProtect)
+        public Models.ProtectedBranch ProtectBranch(BranchProtect branchProtect)
         {
             throw new NotImplementedException();
```
Comment (meziantou) [inline NGitLab.Mock/Clients/ProtectedBranchClient.cs]: I think you can implement the method using something like:

````
var protectedBranch = project.ProtectedBranches.FirstOrDefault(b => b.Name == branchName);
if(protectedBranch == null)
{
    protectedBranch = new();
    project.ProtectedBranches.Add(protectedBranch);
}

// TODO Use actual properties
protectedBranch.Prop = branchProtect.Prop;
protectedBranch.Prop = branchProtect.Prop;
````

File: NGitLab.Mock/Project.cs [hunk 7]
```diff
@@ -146,6 +147,8 @@ public string[] Tags
 
         public ProjectStatistics Statistics { get; set; }
 
+        public ProtectedBranchCollection ProtectedBranches { get; set; }
```
Comment (meziantou) [inline NGitLab.Mock/Project.cs]:
```suggestion
        public ProtectedBranchCollection ProtectedBranches { get; }
```

**Review by meziantou — CHANGES_REQUESTED** [review]

File: NGitLab.Mock/Clients/ProtectedBranchClient.cs [hunk 8]
```diff
@@ -13,24 +14,51 @@ public ProtectedBranchClient(ClientContext context, int projectId)
...
+        public Models.ProtectedBranch ProtectBranch(BranchProtect branchProtect)
         {
-            throw new NotImplementedException();
+            using (Context.BeginOperationScope())
+            {
+                var project = GetProject(_projectId, ProjectPermission.View);
+                var protectedBranch = project.ProtectedBranches.FirstOrDefault(b => b.Name.Equals(branchProtect.BranchName, StringComparison.Ordinal));
+                protectedBranch ??= new();
+
+                project.ProtectedBranches.Add(protectedBranch);
```
Comment (meziantou) [inline NGitLab.Mock/Clients/ProtectedBranchClient.cs]: You must not add the protected branch if it already exists. If the branch is already protected, you should only edit its properties. (see the code in my initial comment)

**Review by meziantou — CHANGES_REQUESTED** [review]

File: NGitLab.Mock/Clients/ProtectedBranchClient.cs (hunk 6 above)
Comment (PrashanthiRamesh) [reply NGitLab.Mock/Clients/ProtectedBranchClient.cs]: Desole. I misunderstood the previous comment. Corrected :)

File: NGitLab.Mock/Clients/ProtectedBranchClient.cs (hunk 8 above)
Comment (PrashanthiRamesh) [reply NGitLab.Mock/Clients/ProtectedBranchClient.cs]: Desole. I misunderstood the previous comment. Corrected :)

**Review by meziantou — APPROVED** [review]

**Pull request merged**